topi/include/topi/nn/pooling.h

```diff
@@ -84,189 +91,170 @@ inline Tensor pool_nchw(const Tensor& x,
-        { dheight, dwidth });
-      }, "tensor", "pool_max");
-  } else if (pool_type == kAvgPool) {
-    auto temp = pad(x, pad_before, pad_after, 0, "pad_temp");
-
-    auto tsum = tvm::compute(
-     { batch, out_height, out_width, channel },
-      [&](Var n, Var h, Var w, Var c) {
-        return tvm::sum(temp(n, h * stride_height + dheight, w * stride_width + dwidth, c),
-        { dheight, dwidth });
-      }, "tensor", "pool_avg");
-
-    return tvm::compute(
-     { batch, out_height, out_width, channel },
-     [&](Var n, Var h, Var w, Var c) {
-       return tsum(n, h, w, c) / (kernel_height * kernel_width);
-      }, "tensor", kElementWise);
-  } else {
-    LOG(ERROR) << "Unrecognized pool_type: " << pool_type;
-    return x;
+inline bool find_height_width(const std::string& layout,
+                              int* height_idx,
+                              int* width_idx) {
+  *height_idx = -1, *width_idx = -1;
+  int curr_idx = 0;
+  for (size_t i = 0; i < layout.size(); ++i) {
+    if ((layout[i] >= 'A' && layout[i] <= 'Z') ||
+        (layout[i] >= 'a' && layout[i] <= 'z')) {
+      if (layout[i] == 'H') {
+        if (*height_idx != -1) return false;
+        *height_idx = curr_idx;
+      } else if (layout[i] == 'W') {
+        if (*width_idx != -1) return false;
+        *width_idx = curr_idx;
+      } else if (layout[i] == 'h' || layout[i] == 'w') {
+        // do not support split on height or width, e.g., NCHW16w
+        return false;
+      }
+      ++curr_idx;
+    }
   }
+  if (*height_idx == -1 || *width_idx == -1) return false;
+  return true;
 }
 
 /*!
-* \brief Perform pooling on data
-*
-* \param x The input tensor in NCHW or NHWC order
+* \brief Perform pooling on height and width dimension of data.
+*        It decides the height and width dimension according to the layout string,
+*        in which 'W' and 'H' means width and height respectively.
+*        Width and height dimension cannot be split.
+*        For example, NCHW, NCHW16c, etc. are valid for pool,
+*        while NCHW16w, NCHW16h are not.
+*        See \a layout for more information of the layout string convention.
+* \param x The input tensor.
 * \param kernel_size Vector of two ints: {kernel_height, kernel_width}
 * \param stride_size Vector of two ints: {stride_height, stride_width}
 * \param padding_size Vector of two ints: {padding_height, padding_width}
 * \param pool_type The type of pooling operator
 * \param ceil_mode Whether to use ceil when calculating the output size
-* \param layout The input layout
-*
-* \return The output tensor in NCHW order
+* \param layout The input layout.
+*        The layout is supposed to be composed of upper cases, lower cases and numbers,
+*        where upper case indicates a dimension and
+*        the corresponding lower case with factor size indicates the split dimension.
+*        For example, NCHW16c can describe a 5-D tensor of
```

Comment:
Do we need to specify the exact split number inside layout? For a 5-D tensor, layout like "NCHWc" should also work and be more general.

Reply:
I prefer to minimize the flexibility of layout as long as it is enough for using. In your case, one has to write "NCHW1c". If there's specific use-case, e.g., lazy infer the size of sub-dim, we can add support later.

Reply:
I think we do need lazy infer while removing data packing. Is this accurate? Since the exact layout of each pooling operator depends on its input shape, though it's still 5-D.

Reply:
Also I'm adding upsampling layout support, since this op is used by another version of SSD. The mechanics is similar to pooling. A more general and flexible implementation will help other similar ops.